### ISSUE TITLE
Edward Roadmap

### ISSUE THREAD

**Issue by dustinvtran** [blei-lab/edward#464]
This is an outline of Edward going forward, at the least for Spring 2017.  Of course, comments are always welcome. I'm happy to change priorities subject to interest. Related issues are added in parentheses.

+ More community support
  + Mailing list
  + Pre-trained models in the model zoo (#28)
  + Repo for sharing Edward content such as papers, talks, and posters
  + Jupyter notebooks for all tutorials (#35)
  + Drafting stages on a fully open source handbook
+ More robustification
  + Integration tests, inference tests (#174)
  + Improved errors and warning messages
  + __Many__ minor bug fixes
+ More automation
  + Automated choice of posterior approximation
  + Automated transformations (#439)
  + Control variates (#4)
  + Convergence diagnostics (#5)
  + Automated hyperparameter tuning
+ More applications & tutorials
  + Sigmoid belief network with Gumbel-Softmax/Concrete distribution ([Jang et al., 2017](https://arxiv.org/abs/1611.01144); [Maddison et al., 2017](https://arxiv.org/abs/1611.00712)) (#299)
  + Large-scale LDA using (vanilla) stochastic variational inference [(Hoffman et al., 2013)](https://arxiv.org/abs/1206.7051) (#423, #463)
  + Generative model applied to ImageNet
  + Hidden Markov model (#450)
  + Other large-scale examples & tutorials (#22, #137)
  + More inference & criticism tutorials (#35)
  + More benchmarks, including a JMLR paper and dedicated webpage
+ More algorithms
  + More GAN algorithms, and in general for implicit probabilistic models (#444)
  + Conjugacy
    + (vanilla) stochastic variational inference [(Hoffman et al., 2013)](https://arxiv.org/abs/1206.7051)
    + Gibbs sampling [(Gelfand and Smith, 1990)](http://www.tandfonline.com/doi/abs/10.1080/01621459.1990.10476213)
  + Expressive variational families ([Rezende and Mohamed, 2015](https://arxiv.org/abs/1505.05770); [Ranganath et al., 2016](https://arxiv.org/abs/1511.02386)) (#7, #284, #447)
  + Gradient-based marginal optimization (Tran et al., 2017)
  + Operator variational inference [(Ranganath et al., 2016)](https://arxiv.org/abs/1610.09033)
  + Inference support for missing data & causality (#399)

Here are features I probably don't have time for. However, they are very much on my mind, and are impactful additions if someone wants to take the helm. (Any of the above is also appreciated!)

+ Better support for Bayesian nonparametric modeling & inference (#183)
+ Better support for undirected models, including potential functions (#443)
+ Sparse random variables (#358)
+ Reinforcement learning applications, interfaced with OpenAI Gym and Universe
+ Dynamic computation graphs
+ Built-in visualization support (#351, #352, #353, #354, #355)
+ Built-in dataset support (#125)
+ Module for contributed models (#380)

**Comment by yyaodong** [comment]
Hi Dustin:

Just wondering if there is any `Control variates (#4)` tricks already implemented in Edward? I noticed there is a branch called `rao-blakcwellization`, not sure if it works properly now.

Thanks.

**Comment by dustinvtran** [comment]
Not at the moment. I was working out RB as you pointed out, which turns out to be quite involved. Control variates are much easier to implement; I haven't looked into it quite yet.

**Comment by ip01** [comment]
Hi Dustin, do you, or perhaps Francisco, or whomever, plan to implement Generalized Reparametrization Gradient in Edward? If yes  - could it be usefull as effective black-box technique for large-scale topic models (Dirichlet G-REP'ed via exp-covariance, Multinomial via Gumbel-Softmax)?

**Comment by dustinvtran** [comment]
I've never experimented with it so I'm not sure. pinging @franrruiz. I do think there's no one-size-fits-all solution—especially for discrete random variables, and whether it score functions+control variates, g-rep, gumbel softmax, or reparameterization gradients. However, it would be certainly useful to have them all to experiment with them.

**Comment by RoyiAvital** [comment]
Is anyone keep working on this?
